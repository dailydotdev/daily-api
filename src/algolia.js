import algoliasearch from 'algoliasearch';
import config from './config';

let client;
let index;

const initAlgolia = (indexName) => {
  client = algoliasearch(config.algolia.app, config.algolia.key);
  index = client.initIndex(`${config.algolia.indexPrefix}_${indexName}`);
};

export const getPostsIndex = () => {
  if (!index) {
    initAlgolia('posts');
  }
  return index;
};

export const trackSearch = (trackingId, ip) => {
  if (client) {
    client.setExtraHeader('X-Algolia-UserToken', trackingId);
    client.setExtraHeader('X-Forwarded-For', ip);
  }
};
