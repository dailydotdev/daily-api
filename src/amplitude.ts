import * as Amplitude from '@amplitude/node';

const amplitude = Amplitude.init(process.env.AMPLITUDE_KEY);

export default amplitude;
