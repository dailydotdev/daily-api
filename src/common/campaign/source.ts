import {
  CampaignSource,
  CampaignState,
  CampaignType,
  Post,
  Source,
  type ConnectionManager,
} from '../../entity';

import type { AuthContext } from '../../Context';
import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import {
  ensureSourcePermissions,
  SourcePermissions,
} from '../../schema/sources';
import { ValidationError } from 'apollo-server-errors';
import { randomUUID } from 'crypto';
import { addDays } from 'date-fns';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { usdToCores } from '../number';
import { systemUser, updateFlagsStatement } from '../utils';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationBuilder } from '../../notifications/builder';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import type { NotificationCampaignSourceContext } from '../../notifications';

export const validateSquadBoostPermissions = async (
  ctx: AuthContext,
  sourceId: string,
) => {
  const source = await ensureSourcePermissions(
    ctx,
    sourceId,
    SourcePermissions.BoostSquad,
  );

  if (source.flags.campaignId) {
    throw new ValidationError('Source already has a campaign');
  }

  return source;
};

export const getSourceTags = async (
  con: ConnectionManager,
  sourceId: string,
): Promise<string[]> => {
  const result = await con.getRepository(Post).query<{ tag: string }[]>(
    `
      WITH recent_posts AS (
        SELECT id, "sharedPostId"
        FROM post
        WHERE "sourceId" = $1
        ORDER BY "createdAt" DESC
        LIMIT 30
      )
      SELECT DISTINCT ps.keyword AS tag
      FROM post_keyword ps
      INNER JOIN recent_posts rp
        ON ps."postId" = rp.id
        OR ps."postId" = rp."sharedPostId"
      WHERE ps.status = 'allow'
      LIMIT 30;
    `,
    [sourceId],
  );

  return result.map(({ tag }) => tag.trim()).filter(Boolean);
};

export const startCampaignSource = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value } = args;
  const source = await validateSquadBoostPermissions(ctx, value);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignSource).save(
      manager.getRepository(CampaignSource).create({
        id,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: source.id,
        state: CampaignState.Active,
        endedAt,
        sourceId: source.id,
        type: CampaignType.Squad,
      }),
    );

    const campaignId = campaign.id;
    const last30tags = await getSourceTags(manager, source.id);
    const finalTags = last30tags.length > 3 ? last30tags : []; // when it is 3 or below, we set global targeting

    await skadiApiClientV2.startCampaign(campaign, finalTags);

    await manager
      .getRepository(Source)
      .update(
        { id: source.id },
        { flags: updateFlagsStatement<Source>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Squad Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignSource = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(CampaignSource)
      .update({ id: campaignId }, { state: CampaignState.Cancelled });

    await manager
      .getRepository(Source)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Source>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Squad Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const generateCampaignSquadNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignSourceContext,
) => {
  const { campaign, source, event, user } = ctx;

  return builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`)
    .avatarSource(source);
};

export const generateCampaignSquadEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(CampaignSource)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  const source = await con
    .getRepository(Source)
    .findOneByOrFail({ id: campaign.sourceId });

  return {
    start_date: formatMailDate(campaign.createdAt),
    end_date: formatMailDate(campaign.endedAt),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    source_image: source.image,
    source_handle: source.handle,
    source_name: source.name,
  };
};
