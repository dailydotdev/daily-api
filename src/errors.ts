import { ApolloError } from 'apollo-server-errors';

export enum UserFailErrorKeys {
  GenericError = 'GENERIC_ERROR',
  MissingFields = 'MISSING_FIELDS',
  UserExists = 'USER_EXISTS',
  UsernameEmailExists = 'USERNAME_EMAIL_EXISTS',
}

export enum UpdateUserFailErrorKeys {
  MissingFields = 'MISSING_FIELDS',
  UserDoesntExist = 'USER_DOESNT_EXIST',
}

export enum SubmissionFailErrorKeys {
  GenericError = 'GENERIC_ERROR',
  Paywall = 'PAYWALL',
  MissingFields = 'MISSING_FIELDS',
  ScoutIsAuthor = 'SCOUT_IS_AUTHOR',
  PostExists = 'POST_EXISTS',
  AuthorBanned = 'AUTHOR_BANNED',
  AccessDenied = 'ACCESS_DENIED',
  LimitReached = 'LIMIT_REACHED',
  InvalidUrl = 'INVALID_URL',
  PostDeleted = 'POST_DELETED',
  ExistsStarted = 'EXISTS_STARTED',
  ExistsAccepted = 'EXISTS_ACCEPTED',
  ExistsRejected = 'EXISTS_REJECTED',
  InviteLimitReached = 'INVITE_LIMIT_REACHED',
  OnboardingTagLimitReached = 'ONBOARDING_TAG_LIMIT_REACHED',
  FeedCountLimitReached = 'FEED_COUNT_LIMIT_REACHED',
  FeedNameRequired = 'FEED_NAME_REQUIRED',
  FeedNameInvalid = 'FEED_NAME_INVALID',
  FeedNameLength = 'FEED_NAME_LENGTH',
}

export const SubmissionFailErrorMessage: Record<
  SubmissionFailErrorKeys,
  string
> = {
  GENERIC_ERROR:
    'Unfortunately there was an error and we were unable to gather the required information from the URL submitted to add',
  PAYWALL:
    'Unfortunately the article submitted is behind a paywall, so we cannot add it to the daily.dev feed.',
  MISSING_FIELDS:
    'Unfortunately we ran into a problem adding this article, our team is looking into it.',
  SCOUT_IS_AUTHOR:
    'You canâ€™t submit your own articles as community picks, please suggest articles by other people.',
  POST_EXISTS: 'This post is already on daily.dev!',
  AUTHOR_BANNED:
    'Unfortunately the article submitted is written by a creator who violated our community guidelines and is banned. We no longer accept submissions from this creator.',
  ACCESS_DENIED:
    'You do not have sufficient permissions and or reputation to submit a community link yet.',
  LIMIT_REACHED:
    'You can only submit 3 links per day and have reached your limit. Please try again tomorrow.',
  INVITE_LIMIT_REACHED: 'You have reached your limit of available invites',
  INVALID_URL:
    'The URL you submitted is not valid, please check and try again.',
  POST_DELETED:
    'This post has previously appeared in the daily.dev feed but was deleted and cannot be added to the feed again.',
  EXISTS_STARTED:
    'This article has already been submitted and is currently being processed.',
  EXISTS_ACCEPTED:
    'This article has already been submitted and is currently being added to the daily.dev feed',
  EXISTS_REJECTED:
    'This article has already been submitted but did not meet our technical requirements, it cannot be submitted again',
  ONBOARDING_TAG_LIMIT_REACHED: 'Tag limit reached',
  FEED_COUNT_LIMIT_REACHED:
    'You have reached maximum number of feeds for your user',
  FEED_NAME_REQUIRED: 'Feed name is required',
  FEED_NAME_INVALID: 'Feed name should not contain special characters',
  FEED_NAME_LENGTH: 'Feed name is too long',
};

export enum SourceRequestErrorKeys {
  AccessDenied = 'ACCESS_DENIED',
  SquadIneligible = 'SQUAD_INELIGIBLE',
}

export const SourceRequestErrorMessage: Record<SourceRequestErrorKeys, string> =
  {
    [SourceRequestErrorKeys.AccessDenied]:
      'You do not have sufficient permissions and or reputation to submit a source request yet.',
    [SourceRequestErrorKeys.SquadIneligible]:
      'Squad has not been approved yet of becoming public',
  };

export class NotFoundError extends ApolloError {
  constructor(message: string) {
    super(message, 'NOT_FOUND');

    Object.defineProperty(this, 'name', { value: 'NotFoundError' });
  }
}

export enum TypeOrmError {
  NULL_VIOLATION = '23502',
  FOREIGN_KEY = '23503',
  DUPLICATE_ENTRY = '23505',
  USER_CONSTRAINT = 'FK_dce2a8927967051c447ae10bc8b',
}

export enum SourcePermissionErrorKeys {
  InviteInvalid = 'SOURCE_PERMISSION_INVITE_INVALID',
}

// Return 409 HTTP status code
export class ConflictError extends ApolloError {
  constructor(message) {
    super(message, 'CONFLICT');

    Object.defineProperty(this, 'name', { value: 'ConflictError' });
  }
}
