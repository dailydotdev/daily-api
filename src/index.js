import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import KoaPinoLogger from 'koa-pino-logger';
import Router from 'koa-router';
import etag from 'koa-etag';


import config from './config';
// import compress from './middlewares/compress';
import errorHandler from './middlewares/errorHandler';
import logger from './logger';

import health from './routes/health';
import sources from './routes/sources';
import posts from './routes/posts';
import publications from './routes/publications';
import tweet from './routes/tweet';
import settings from './routes/settings';
import feeds from './routes/feeds';
import notifications from './routes/notifications';
import tags from './routes/tags';

const app = new Koa();

app.keys = [config.cookies.secret];

app.proxy = true;

// TODO: disabled due to performance issues
// app.use(compress());

app.use(bodyParser());
app.use(KoaPinoLogger({ logger, useLevel: 'debug' }));
app.use(errorHandler());
app.use(etag());

// Machine-to-machine authentication
app.use((ctx, next) => {
  if (ctx.request.get('authorization') === `Service ${config.accessSecret}`
    && ctx.request.get('user-id') && ctx.request.get('logged-in')) {
    // eslint-disable-next-line
    ctx.state = {
      user: {
        userId: ctx.request.get('user-id'),
      },
      service: true,
    };
  } else {
    delete ctx.request.headers['user-id'];
    delete ctx.request.headers['logged-in'];
  }
  return next();
});

const router = new Router({
  prefix: '/v1',
});

router.use(feeds.routes(), feeds.allowedMethods());
router.use(sources.routes(), sources.allowedMethods());
router.use(posts.routes(), posts.allowedMethods());
router.use(publications.routes(), publications.allowedMethods());
router.use(tweet.routes(), tweet.allowedMethods());
router.use(settings.routes(), settings.allowedMethods());
router.use(notifications.routes(), notifications.allowedMethods());
router.use(tags.routes(), tags.allowedMethods());

app.use(router.routes(), router.allowedMethods());
app.use(health.routes(), health.allowedMethods());

export default app;
