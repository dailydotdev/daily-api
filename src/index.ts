import 'reflect-metadata';
import fastify, {
  FastifyRequest,
  FastifyInstance,
  FastifyError,
} from 'fastify';
import fastifyRawBody from 'fastify-raw-body';
import helmet from '@fastify/helmet';
import cors from '@fastify/cors';
import mercurius, { MercuriusError } from 'mercurius';
import MercuriusGQLUpload from 'mercurius-upload';
import MercuriusCache from 'mercurius-cache';
import { NoSchemaIntrospectionCustomRule } from 'graphql';
// import fastifyWebsocket from '@fastify/websocket';

import './config';

import tracking from './tracking';
import auth from './auth';
import compatibility from './compatibility';
import routes from './routes';
import { Context } from './Context';
import { schema } from './graphql';
import createOrGetConnection from './db';
import { stringifyHealthCheck } from './common';
import { GraphQLError } from 'graphql';
import cookie, { FastifyCookieOptions } from '@fastify/cookie';
import { getSubscriptionSettings } from './subscription';
import { ioRedisPool } from './redis';
import { loadFeatures } from './growthbook';
import { runInRootSpan } from './telemetry/opentelemetry';
import { loggerConfig } from './logger';

type Mutable<Type> = {
  -readonly [Key in keyof Type]: Type[Key];
};

const userExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined => (ctx.userId ? `user:${ctx.userId}` : undefined);

const trackingExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined =>
  ctx.trackingId ? `tracking:${ctx.trackingId}` : undefined;

// readiness probe is set failureThreshold: 2, periodSeconds: 2 (4s) + small delay
const GRACEFUL_DELAY = 2 * 2 * 1000 + 5000;

export default async function app(
  contextFn?: (request: FastifyRequest) => Context,
): Promise<FastifyInstance> {
  let isTerminating = false;
  const isProd = process.env.NODE_ENV === 'production';
  const connection = await runInRootSpan(
    'createOrGetConnection',
    createOrGetConnection,
  );

  const app = fastify({
    logger: loggerConfig,
    disableRequestLogging: true,
    trustProxy: true,
  });

  app.log.info('loading features');
  await loadFeatures(app.log);

  const gracefulShutdown = () => {
    app.log.info('starting termination');
    isTerminating = true;
    setTimeout(async () => {
      await app.close();
      await connection.destroy();
      await ioRedisPool.end();
      process.exit();
    }, GRACEFUL_DELAY);
  };
  process.on('SIGINT', gracefulShutdown);
  process.on('SIGTERM', gracefulShutdown);

  app.register(helmet);
  app.register(cors, {
    origin: isProd ? /daily\.dev$/ : true,
    credentials: true,
  });
  app.register(cookie, {
    secret: process.env.COOKIES_KEY,
  }) as FastifyCookieOptions;
  app.register(auth, { secret: process.env.ACCESS_SECRET });
  app.register(tracking);
  app.register(fastifyRawBody, {
    field: 'rawBody',
    global: false,
    encoding: 'utf8',
    runFirst: true,
  });

  app.setErrorHandler((err, req, res) => {
    req.log.error({ err }, err.message);
    res.code(500).send({ statusCode: 500, error: 'Internal Server Error' });
  });

  app.get('/health', (req, res) => {
    res.type('application/health+json');
    if (isTerminating) {
      res.status(500).send(stringifyHealthCheck({ status: 'terminating' }));
    } else {
      res.send(stringifyHealthCheck({ status: 'ok' }));
    }
  });

  app.get('/liveness', (req, res) => {
    res.type('application/health+json');
    res.send(stringifyHealthCheck({ status: 'ok' }));
  });

  // app.register(fastifyWebsocket, {
  //   options: {
  //     maxPayload: 1048576,
  //     verifyClient: (info, next) => next(true),
  //   },
  // });

  app.register(MercuriusGQLUpload, {
    maxFileSize: 1024 * 1024 * 20,
    maxFiles: 1,
  });

  app.register(mercurius, {
    schema,
    context:
      contextFn ?? ((request): Context => new Context(request, connection)),
    queryDepth: 10,
    subscription: getSubscriptionSettings(connection),
    // Disable GraphQL introspection in production
    graphiql: !isProd,
    validationRules: isProd && [NoSchemaIntrospectionCustomRule],
    errorFormatter(execution, ctx) {
      if (execution.errors?.length > 0) {
        const flatErrors = execution.errors.flatMap<GraphQLError>((error) => {
          if (error.originalError.name === 'FastifyError') {
            return (
              (error.originalError as MercuriusError<GraphQLError>).errors ||
              error
            );
          }
          return error;
        });
        return {
          statusCode: 200,
          response: {
            data: execution.data,
            errors: flatErrors.map((error): GraphQLError => {
              const newError = error as Mutable<GraphQLError>;
              if (!error.originalError) {
                newError.extensions = {
                  code: 'GRAPHQL_VALIDATION_FAILED',
                };
              } else if (error.originalError?.name === 'EntityNotFoundError') {
                newError.message = 'Entity not found';
                newError.extensions = {
                  code: 'NOT_FOUND',
                };
              } else if (
                (error.originalError as FastifyError)?.code ===
                'MER_ERR_GQL_PERSISTED_QUERY_NOT_FOUND'
              ) {
                app.log.debug(
                  { body: ctx?.reply?.request?.body },
                  'unknown query',
                );
              } else if (!error.extensions?.code) {
                app.log.warn(
                  {
                    err: error.originalError,
                    body: ctx?.reply?.request?.body,
                  },
                  'unexpected graphql error',
                );
                newError.message = 'Unexpected error';
                newError.extensions = {
                  code: 'UNEXPECTED',
                };
              }
              if (isProd) {
                newError.originalError = undefined;
              }
              return newError;
            }),
          },
        };
      }
      return {
        statusCode: 200,
        response: execution,
      };
    },
  });

  if (isProd) {
    app.register(MercuriusCache, {
      ttl: 10,
      policy: {
        Query: {
          searchBookmarksSuggestions: {
            extendKey: userExtendKey,
          },
          searchBookmarks: {
            extendKey: userExtendKey,
          },
          // anonymousFeed: {
          //   extendKey: userExtendKey,
          // },
          // feed: {
          //   extendKey: userExtendKey,
          // },
          sourceFeed: {
            extendKey: userExtendKey,
          },
          tagFeed: {
            extendKey: userExtendKey,
          },
          keywordFeed: {
            extendKey: userExtendKey,
          },
          searchPostSuggestions: {
            extendKey: userExtendKey,
          },
          searchPosts: {
            extendKey: userExtendKey,
          },
          authorFeed: {
            extendKey: userExtendKey,
          },
          mostUpvotedFeed: {
            extendKey: userExtendKey,
          },
          mostDiscussedFeed: {
            extendKey: userExtendKey,
          },
          randomTrendingPosts: {
            extendKey: userExtendKey,
          },
          randomSimilarPostsByTags: {
            extendKey: userExtendKey,
          },
          randomDiscussedPosts: {
            extendKey: userExtendKey,
          },
          tagsCategories: true,
          advancedSettings: {
            extendKey: trackingExtendKey,
          },
          banner: true,
          postByUrl: {
            extendKey: userExtendKey,
          },
          post: {
            extendKey: userExtendKey,
          },
          postUpvotes: true,
          searchTags: true,
          userReadingRankHistory: true,
          userReadHistory: true,
        },
      },
    });
  }

  app.register(compatibility, { prefix: '/v1' });
  app.register(routes, { prefix: '/' });

  return app;
}
