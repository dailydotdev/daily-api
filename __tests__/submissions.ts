import {
  ArticlePost,
  Source,
  User,
  Submission,
  SubmissionStatus,
} from '../src/entity';
import {
  disposeGraphQLTesting,
  GraphQLTestClient,
  GraphQLTestingState,
  initializeGraphQLTesting,
  MockContext,
  saveFixtures,
  testMutationErrorCode,
} from './helpers';
import { DEFAULT_SUBMISSION_LIMIT } from '../src/schema/submissions';
import { subDays } from 'date-fns';
import { sourcesFixture } from './fixture/source';
import { SubmissionFailErrorMessage } from '../src/errors';
import { DataSource } from 'typeorm';
import createOrGetConnection from '../src/db';
import { utcToZonedTime, zonedTimeToUtc } from 'date-fns-tz';

let con: DataSource;
let state: GraphQLTestingState;
let client: GraphQLTestClient;
let loggedUser: string = null;

beforeAll(async () => {
  con = await createOrGetConnection();
  state = await initializeGraphQLTesting(
    () => new MockContext(con, loggedUser),
  );
  client = state.client;
});

beforeEach(async () => {
  loggedUser = null;
  await con.getRepository(User).save({
    id: '1',
    name: 'Lee',
    image: 'https://daily.dev/lee.jpg',
    reputation: 250,
  });
});

afterAll(() => disposeGraphQLTesting(state));

describe('query submissionAvailability', () => {
  const QUERY = `
    query SubmissionAvailability {
      submissionAvailability {
        hasAccess
        limit
        todaySubmissionsCount
      }
    }
  `;

  it('should return default values if not logged in', async () => {
    loggedUser = '0';

    const res = await client.query(QUERY);
    const limit = parseInt(
      process.env.SCOUT_SUBMISSION_LIMIT || DEFAULT_SUBMISSION_LIMIT,
    );
    expect(res.errors).toBeFalsy();
    expect(res.data.submissionAvailability.limit).toEqual(limit);
    expect(res.data.submissionAvailability.hasAccess).toEqual(false);
    expect(res.data.submissionAvailability.todaySubmissionsCount).toEqual(0);
  });

  it('should return submissions count today, limit, and if has access', async () => {
    loggedUser = '1';
    const repo = con.getRepository(Submission);
    await repo.save([
      { url: 'http://abc.com/1', userId: '1' },
      {
        url: 'http://abc.com/2',
        userId: '1',
        createdAt: subDays(new Date(), 1),
      },
    ]);
    const res = await client.query(QUERY);
    const limit = parseInt(
      process.env.SCOUT_SUBMISSION_LIMIT || DEFAULT_SUBMISSION_LIMIT,
    );
    expect(res.errors).toBeFalsy();
    expect(res.data.submissionAvailability.limit).toEqual(limit);
    expect(res.data.submissionAvailability.hasAccess).toEqual(true);
    expect(res.data.submissionAvailability.todaySubmissionsCount).toEqual(1);
  });

  it('should return submissions count today, limit, and if has access in consideration of timezone', async () => {
    loggedUser = '1';
    await con.getRepository(User).save({
      id: '2',
      name: 'Hansel',
      image: 'https://daily.dev/hansel.jpg',
      reputation: 250,
    });
    await con.getRepository(User).save({
      id: '3',
      name: 'Solevilla',
      image: 'https://daily.dev/solevilla.jpg',
      reputation: 250,
    });
    await con
      .getRepository(User)
      .update({ id: '1' }, { timezone: 'Europe/Oslo' });
    const repo = con.getRepository(Submission);
    await repo.save([
      { url: 'http://abc.com/1', userId: '2' },
      { url: 'http://abc.com/2', userId: '3' },

      // 20:00 at UTC is 21:00 at Europe/Oslo (22:00 during summer)
      // So the submission should be counted as yesterday
      {
        url: 'http://abc.com/3',
        userId: '1',
        createdAt: subDays(new Date().setHours(20), 1),
      },

      // 23:00 at UTC is 01:00 at Europe/Oslo (02:00 during summer)
      // So the submission should be counted as the next day
      {
        url: 'http://abc.com/4',
        userId: '1',
        createdAt: subDays(new Date().setHours(23), 1),
      },
      {
        url: 'http://abc.com/4.midnight',
        userId: '1',
        createdAt: zonedTimeToUtc(
          utcToZonedTime(new Date().setHours(23), 'Europe/Oslo'),
          'Europe/Oslo',
        ),
      },
      {
        url: 'http://abc.com/5',
        userId: '1',
      },
    ]);
    const res = await client.query(QUERY);
    const limit = parseInt(
      process.env.SCOUT_SUBMISSION_LIMIT || DEFAULT_SUBMISSION_LIMIT,
    );
    expect(res.errors).toBeFalsy();
    expect(res.data.submissionAvailability.limit).toEqual(limit);
    expect(res.data.submissionAvailability.hasAccess).toEqual(true);
    expect(res.data.submissionAvailability.todaySubmissionsCount).toEqual(2);
  });
});

describe('mutation submitArticle', () => {
  const MUTATION = `
    mutation SubmitArticle($url: String!) {
      submitArticle(url: $url) {
        result
        reason
        post {
          id
        }
        submission {
          id
          status
          userId
        }
      }
    }
  `;

  it('should not authorize when not logged in', () =>
    testMutationErrorCode(
      client,
      { mutation: MUTATION, variables: { url: 'http://abc.com' } },
      'UNAUTHENTICATED',
    ));

  it('should invalidate if the url was requested already', async () => {
    loggedUser = '1';
    const request = 'https://abc.com/article';
    const repo = con.getRepository(Submission);
    await repo.save(repo.create({ url: request, userId: loggedUser }));

    const res = await client.mutate(MUTATION, { variables: { url: request } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toEqual({
      submitArticle: {
        result: 'rejected',
        reason: SubmissionFailErrorMessage.EXISTS_STARTED,
        post: null,
        submission: null,
      },
    });
  });

  it('should invalidate if the user has reached limit', async () => {
    loggedUser = '1';
    const request = 'https://abc.com/article';
    const repo = con.getRepository(Submission);
    await repo.save(repo.create({ url: `${request}1`, userId: loggedUser }));
    await repo.save(repo.create({ url: `${request}2`, userId: loggedUser }));
    await repo.save(repo.create({ url: `${request}3`, userId: loggedUser }));

    const res = await client.mutate(MUTATION, { variables: { url: request } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toEqual({
      submitArticle: {
        result: 'rejected',
        reason: SubmissionFailErrorMessage.LIMIT_REACHED,
        post: null,
        submission: null,
      },
    });
  });

  it('should invalidate if the user is ineligible for submission', async () => {
    loggedUser = '1';
    await con.getRepository(User).update({ id: '1' }, { reputation: 10 });
    const request = 'https://abc.com/article';
    const res = await client.mutate(MUTATION, { variables: { url: request } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toEqual({
      submitArticle: {
        result: 'rejected',
        reason: SubmissionFailErrorMessage.ACCESS_DENIED,
        post: null,
        submission: null,
      },
    });
  });

  it('should reject if the post already exists', async () => {
    loggedUser = '1';
    const request = 'http://p1.com';
    await saveFixtures(con, Source, sourcesFixture);
    await saveFixtures(con, ArticlePost, [
      {
        id: 'p1',
        shortId: 'sp1',
        title: 'Post 1',
        url: request,
        canonicalUrl: request,
        score: 0,
        sourceId: 'a',
        createdAt: new Date('2021-09-22T07:15:51.247Z'),
        tagsStr: 'javascript,webdev',
      },
    ]);

    const res = await client.mutate(MUTATION, { variables: { url: request } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toEqual({
      submitArticle: {
        result: 'exists',
        reason: null,
        post: {
          id: 'p1',
        },
        submission: null,
      },
    });
  });

  it('should reject if the post was deleted', async () => {
    loggedUser = '1';
    const request = 'http://p8.com';
    await saveFixtures(con, Source, sourcesFixture);
    await saveFixtures(con, ArticlePost, [
      {
        id: 'pdeleted',
        shortId: 'spdeleted',
        title: 'PDeleted',
        url: request,
        canonicalUrl: request,
        score: 0,
        sourceId: 'a',
        createdAt: new Date('2021-09-22T07:15:51.247Z'),
        tagsStr: 'javascript,webdev',
        deleted: true,
      },
    ]);

    const res = await client.mutate(MUTATION, { variables: { url: request } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toEqual({
      submitArticle: {
        result: 'rejected',
        reason: SubmissionFailErrorMessage.POST_DELETED,
        post: null,
        submission: null,
      },
    });
  });

  it('should not allow invalid urls', async () => {
    loggedUser = '1';
    const request = 'test/sample/url';
    const res = await client.mutate(MUTATION, { variables: { url: request } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toEqual({
      submitArticle: {
        result: 'rejected',
        reason: SubmissionFailErrorMessage.INVALID_URL,
        post: null,
        submission: null,
      },
    });
  });

  it('should create a submission entity if the url is valid', async () => {
    loggedUser = '1';
    const request = 'https://daily.dev/amazing/article';
    const res = await client.mutate(MUTATION, { variables: { url: request } });
    expect(res.errors).toBeFalsy();
    const submission = await con
      .getRepository(Submission)
      .findOneBy({ url: request });
    expect(submission.status).toEqual(SubmissionStatus.Started);
    expect(res.data).toMatchSnapshot({
      submitArticle: {
        submission: {
          id: expect.any(String),
        },
      },
    });
  });
});
