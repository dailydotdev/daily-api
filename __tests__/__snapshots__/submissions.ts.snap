// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation submitArticle should create a submission entity if the url is valid 1`] = `
Object {
  "submitArticle": Object {
    "post": null,
    "reason": null,
    "result": "succeed",
    "submission": Object {
      "id": Any<String>,
      "status": "STARTED",
      "userId": "1",
    },
  },
}
`;
